xonsh/environ.py

```diff
@@ -1007,6 +1013,36 @@ def _collapsed_pwd():
     base = [i[0] if ix != l-1 else i for ix,i in enumerate(pwd) if len(i) > 0]
     return leader + sep.join(base)
 
+def _dynamically_collapsed_pwd():
+    sep = get_sep()
+    pwd = _replace_home_cwd().split(sep)
+    cols, _ = shutil.get_terminal_size()
+    targetWidthRaw = builtins.__xonsh_env__['DYNAMIC_CWD_WIDTH']
+
+    if (targetWidthRaw[-1] == '%'):
+        targetWidth = math.floor(float(targetWidthRaw[:-1]) / 100 * cols)
+    else:
+        targetWidth = int(targetWidthRaw)
+
+    last = pwd.pop()
+    remaining_space = targetWidth - len(last) 
+    # Reserve space for separators
+    remaining_space_for_text = remaining_space - len(pwd) 
+
+    parts = []
+    for i  in range(len(pwd)):
+        part = pwd[i]
+        part_len = min(len(part), max(1, math.floor(remaining_space_for_text / (len(pwd) - i))))
```

Comment:
Thank you for the tip, it's really appreciated since I'm quite new to python. (I never used it in large scale project only in algorithmic challenges)